### ISSUE TITLE
Key naming strategies (under_scores, CamelCase...)

### ISSUE THREAD

**Issue by crumbpicker** [salat/salat#107]
Some JSON serializers allow to use custom key naming strategies.

Jackson for instance implements the CAMEL_CASE_TO_LOWER_CASE_WITH_UNDERSCORES strategy as a builtin (http://jackson.codehaus.org/1.9.9/javadoc/org/codehaus/jackson/map/PropertyNamingStrategy.LowerCaseWithUnderscoresStrategy.html).

Here is a similar functionality implemented in the GSON library:
https://sites.google.com/site/gson/gson-user-guide#TOC-JSON-Field-Naming-Support
http://google-gson.googlecode.com/svn/trunk/gson/docs/javadocs/com/google/gson/FieldNamingPolicy.html#LOWER_CASE_WITH_UNDERSCORES

It would be really great if Salat could support such kinds of strategies.

**Comment by noahlz** [comment]
(Triaging old issues)

Is this a request for just key fields (`_id`) or for any given field?

If the request is to convert any given case class field to a different case format, it would have to be something like `com/novus/salat/Field.scala` but with quite a bit more reflection code.

I know that this is possible - as mentioned above, Jackson and GSON do something like this, and the awesome `enumeratum` library also has class-to-string naming policies.

@rcaloras @crumbpicker - to say that it's "been a while" would be a gross understatement, but if one of you wants to take a crack at this off the `1.10.x-branch` branch, go for it!

**Comment by crumbpicker** [comment]
Hi @noahlz !

I'm not more a Salat user I moved to reactivemongo and Play Json (because my main project is now using the Play Framework).
They are working on the same issue right now:
https://github.com/playframework/playframework/issues/2333
They are using Scala macros instead of usual reflection approaches.

In reactivemongo they use a @Key annotation with some kind of reflection.
https://github.com/ReactiveMongo/ReactiveMongo/pull/140

I hope it could help !

**Comment by noahlz** [comment]
Interesting - thank you!

Are you using [reactivemongo-play-json](https://github.com/reactivemongo/reactivemongo-play-json)?

**Comment by crumbpicker** [comment]
Yes I'm using the driver, the Play plugin and some home-made classes for the DAO part.
The reactivemongo extensions (https://github.com/fehmicansaglam/reactivemongo-extensions) seem to be abandoned.
